pkg/manager/member/restarter.go

```diff
@@ -0,0 +1,151 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package member
+
+import (
+	"fmt"
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	"github.com/pingcap/tidb-operator/pkg/label"
+	operatorUtil "github.com/pingcap/tidb-operator/pkg/util"
+	apps "k8s.io/api/apps/v1"
+	core "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	appslisters "k8s.io/client-go/listers/apps/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+)
+
+const (
+	emptyRestartPodList = "empty restart pod List"
+)
+
+type Restarter interface {
+	Sync(tc *v1alpha1.TidbCluster, memberType v1alpha1.MemberType) error
+}
+
+type GeneralRestarter struct {
+	kubeCli   kubernetes.Interface
+	podLister corelisters.PodLister
+	stsLister appslisters.StatefulSetLister
+}
+
+func NewGeneralRestarter(kubeCli kubernetes.Interface, podLister corelisters.PodLister, stsLister appslisters.StatefulSetLister) *GeneralRestarter {
+	return &GeneralRestarter{kubeCli: kubeCli, podLister: podLister, stsLister: stsLister}
+}
+
+func (gr *GeneralRestarter) Sync(tc *v1alpha1.TidbCluster, memberType v1alpha1.MemberType) error {
+	sts := &apps.StatefulSet{}
+	err := gr.syncRestartStatus(tc, memberType, sts)
+	if err != nil {
+		return err
+	}
+	return gr.sync(tc, memberType, sts)
+}
+
+// syncRestartStatus would check the statefulset of tc for each component whether they were restarting.
+// If they are, it would check whether the restarting is finished and update the sts annotations.
+func (gr *GeneralRestarter) syncRestartStatus(tc *v1alpha1.TidbCluster, memberType v1alpha1.MemberType, sts *apps.StatefulSet) error {
+	namespace := tc.Namespace
+	stsName := operatorUtil.GetStatefulSetName(tc, memberType)
+	sts, err := gr.stsLister.StatefulSets(namespace).Get(stsName)
+	if err != nil {
+		return err
+	}
+	podName, existed := sts.Annotations[label.AnnPodRestarting]
+	if !existed {
+		return nil
+	}
+	pod, err := gr.podLister.Pods(namespace).Get(podName)
+	if err != nil {
+		return err
+	}
+	if _, existed = pod.Annotations[label.AnnPodDeferDeleting]; existed {
+		return nil
+	}
+	delete(sts.Annotations, label.AnnPodRestarting)
+	_, err = gr.kubeCli.AppsV1().StatefulSets(namespace).Update(sts)
+	if err != nil {
+		return err
+	}
+	return controller.RequeueErrorf("tc[%s/%s]'s pod[%s/%s] is restarted,requeue", namespace, tc.Name, namespace, podName)
+}
+
+// sync func would continue to try deleting the pod which is during restarting or pop one new pod with AnnPodDeferDeleting
+// to restart
+func (gr *GeneralRestarter) sync(tc *v1alpha1.TidbCluster, memberType v1alpha1.MemberType, sts *apps.StatefulSet) error {
+	podName, existed := sts.Annotations[label.AnnPodRestarting]
+	if existed {
+		return gr.restart(tc, memberType, podName)
```

Comment:
If the `podName` is not exist, `restart` will return a error and the restarting lock will never get cleaned